### PR TITLE
Mac: File modified notifications

### PR DESCRIPTION
**Pull request by wilbaker**
Part of the work required for #152 

These changes add file modified notification handling to kernel and user mode code.

Additional functional tests have also been enabled:

```
Test Run Summary
  Overall result: Warning
  Test Count: 143, Passed: 141, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 2
    Skipped Tests - Ignored: 2, Explicit: 0, Other: 0
  Start time: 2018-08-13 22:00:15Z
    End time: 2018-08-13 22:11:28Z
    Duration: 672.380 seconds
```

### CODE REVIEW

File: GVFS/GVFS.Common/FileBasedCollection.cs [hunk 1]
```diff
@@ -390,6 +390,7 @@ private void WriteToDisk(string value)
                 throw new InvalidOperationException(nameof(this.WriteToDisk) + " requires that collectionAppendsDirectlyToFile be true");
             }
 
+            // TODO(Mac): Decide if we want to stick with \r\n on all platforms, move them all to \n, or use a mix
```
Comment (sanoursa) [inline GVFS/GVFS.Common/FileBasedCollection.cs]: Do we need to keep this TODO? As far as I know, both Windows and Mac are happy with this format as is.

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs [hunk 2]
```diff
@@ -299,7 +300,8 @@ public void OverwrittenFileAddedToSparseCheckoutAndSkipWorktreeBitCleared()
 
             fileToOverwriteVirtualPath.ShouldBeAFile(this.fileSystem).WithContents(testContents);
 
-            GVFSHelpers.ModifiedPathsShouldContain(this.fileSystem, this.Enlistment.DotGVFSRoot, fileToOverwriteEntry + Environment.NewLine);
+            // TODO(Mac): Switch from "\r\n" to Environment.NewLine or "\n" depending on how FileBasedCollection is updated
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs]: Same here

File: GVFS/GVFS.Common/FileBasedCollection.cs (hunk 1 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/FileBasedCollection.cs]: @sanoursa Good point, both platforms appear to be happy with it so we can take it out

File: GVFS/GVFS.FunctionalTests/Tools/GVFSFunctionalTestEnlistment.cs [hunk 3]
```diff
@@ -233,9 +234,30 @@ public void UnmountAndDeleteAll()
...
 
-        public string GetVirtualPathTo(params string[] pathInRepo)
+        public string GetVirtualPathTo(string path)
         {
-            return Path.Combine(this.RepoRoot, Path.Combine(pathInRepo));
+            string cleanedPath;
+            switch (Path.DirectorySeparatorChar)
+            {
+                case '/':
+                    cleanedPath = path.Replace('\\', Path.DirectorySeparatorChar);
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/Tools/GVFSFunctionalTestEnlistment.cs]: This line is a bit worrisome - it seems like it could hide tests that are still assuming Windows slashes, rather than force us to clean up the tests.

I would expect this method to convert from Git paths to file system paths, but not to cover over platform-related differences. So to me, that means that it should do a conversion on Windows, and noop on Mac. Thoughts?

File: GVFS/GVFS.Common/FileBasedCollection.cs (hunk 1 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/FileBasedCollection.cs]: Resolved

File: ProjFS.Mac/PrjFSLib.Mac.Managed/Interop/PrjFSLib.cs [hunk 4]
```diff
@@ -25,20 +25,17 @@ internal static class PrjFSLib
...
-        
             [MarshalAs(UnmanagedType.LPArray, SizeConst = PlaceholderIdLength)]
             byte[] providerId,
-
             [MarshalAs(UnmanagedType.LPArray, SizeConst = PlaceholderIdLength)]
             byte[] contentId,
-        
             ulong fileSize,
-            UInt16 fileMode);
+            ushort fileMode);
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSLib.Mac.Managed/Interop/PrjFSLib.cs]: I'm curious, why did you change this type?

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs]: Resolved

File: GVFS/GVFS.Common/FileBasedCollection.cs (hunk 1 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/FileBasedCollection.cs]: Resolved

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs]: Resolved

File: ProjFS.Mac/PrjFSLib.Mac.Managed/Interop/PrjFSLib.cs (hunk 4 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSLib.Mac.Managed/Interop/PrjFSLib.cs]: @sanoursa I had to update these types after enabling StyleCop for the project

File: GVFS/GVFS.FunctionalTests/Tools/GVFSFunctionalTestEnlistment.cs (hunk 3 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tools/GVFSFunctionalTestEnlistment.cs]: @sanoursa I went back and forth on this as well.  I agree with your assessment that we might miss cleaning up tests with this change.  I'll switch this to just convert Git path to system path.

File: ProjFS.Mac/PrjFSLib.Mac.Managed/VirtualizationInstance.cs [hunk 5]
```diff
@@ -18,8 +33,12 @@ public class VirtualizationInstance
             {
                 OnEnumerateDirectory = this.OnEnumerateDirectory,
                 OnGetFileStream = this.OnGetFileStream,
+
+                // Note: We must use this.notifyOperationCallback and not this.OnNotifyOperation
+                // to ensure we don't create a temporary delegate that will be garbage collected
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSLib.Mac.Managed/VirtualizationInstance.cs]: It would be nice to make this more explicit in the code itself. Perhaps something like this?

```
OnNotifyOperation = (this.preventGCOnNotifyOperationDelegate = new NotifyOperationCallback(this.OnNotifyOperation)),
```

The root cause of the issue is that there is an implicit delegate allocation, so let's make at least that much of it explicit in the code. I think it would also be clearer to store that instance here, rather than in the constructor.

**Comment by sanoursa** [comment]
I had tested your last round of changes and it all worked great. I was able to write changes into a newly hydrated file, and `git status` picked up those changes.

File: ProjFS.Mac/PrjFSLib.Mac.Managed/Interop/PrjFSLib.cs (hunk 4 above)
Comment (sanoursa) [reply ProjFS.Mac/PrjFSLib.Mac.Managed/Interop/PrjFSLib.cs]: :+1:

File: ProjFS.Mac/PrjFSLib.Mac.Managed/VirtualizationInstance.cs (hunk 5 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSLib.Mac.Managed/VirtualizationInstance.cs]: @sanoursa 

> I think it would also be clearer to store that instance here, rather than in the constructor.

That sounds good to me (note that `preventGCOnNotifyOperationDelegate` will no longer be able to be `readonly` with this change).
